utcaApp/src/rtmlamp.cpp

```diff
@@ -12,88 +12,61 @@ namespace {
-              {"SP_EFF", p_eff_sp},
-          }),
+              {"AMP_STATUS", p_read_only},
+              {"AMP_STATUS_LATCH", p_read_only},
+              {"AMP_EN", p_decoder_controller},
+              {"MODE", p_mode},
+              {"TRIG_EN", p_decoder_controller},
+              {"RST_LATCH", p_decoder_controller},
+              {"PI_KP", p_decoder_controller},
+              {"PI_TI", p_decoder_controller},
+              {"CNT", p_decoder_controller},
+              {"LIMIT_A", p_decoder_controller},
+              {"LIMIT_B", p_decoder_controller},
+              {"PI_SP", p_decoder_controller},
+              {"DAC", p_decoder_controller},
+              {"ADC_INST", p_read_only},
+              {"DAC_EFF", p_read_only},
+              {"SP_EFF", p_read_only},
+          },
+          &ctl),
       dec(bars),
       ctl(bars)
     {
         auto v = find_device(port_number);
         dec.set_devinfo(v);
         ctl.set_devinfo(v);
 
-        /* XXX: initialize this one specifically to avoid an exception */
-        for (unsigned addr = 0; addr < number_of_channels; addr++) {
-            setIntegerParam(addr, p_opmode, 0);
-        }
-
-        write_only = {p_reset_latch};
+        decoder_controller.insert(p_mode);
 
         read_parameters();
     }
 
-    asynStatus writeInt32Impl(asynUser *pasynUser, const int function, const int addr, epicsInt32 value)
+    asynStatus readEnum(asynUser *pasynUser, char *strings[], int values[],
+        int severities[], size_t, size_t *nIn)
     {
-        setIntegerParam(addr, function, value);
-
-        ctl.channel = addr;
-
-        auto write_param = [addr, this](auto fn, auto &ctl_value) {
-            epicsInt32 tmp;
-            /* we expect success, unless the parameter hasn't been initialized yet */
-            if (getIntegerParam(addr, fn, &tmp) == asynSuccess)
-                ctl_value = tmp;
-            else
-                ctl_value = 0;
-        };
-
-        write_param(p_pwrstate, ctl.amp_enable);
-        write_param(p_opmode, ctl.mode_numeric);
-        write_param(p_loopkp, ctl.pi_kp);
-        write_param(p_loopti, ctl.pi_ti);
-        write_param(p_testwaveperiod, ctl.cnt);
-        write_param(p_testlim_a, ctl.limit_a);
-        write_param(p_testlim_b, ctl.limit_b);
-        write_param(p_pi_sp, ctl.pi_sp);
-        write_param(p_dac, ctl.dac);
-        /* XXX: if someone writes 1 into trigen and it's not supported by the hardware,
-         * we'll error out continuously until zeroed-out again */
-        write_param(p_trigen, ctl.trigger_enable);
-        write_param(p_reset_latch, ctl.reset_latch);
+        const int function = pasynUser->reason;
+        if (function == p_mode)
+            return fill_enum(strings, values, severities, nIn, lamp::mode_list);
+        return asynError;
+    }
 
+    asynStatus writeInt32Impl(asynUser *pasynUser, const int, const int, epicsInt32)
+    {
         return write_params(pasynUser, ctl);
     }
```

Comment:
Just to be sure I'm understanding why this is still here: after GIE new changes to the module, we will now require to implement this, and this will be done in this PR. Is that it?

Reply:
This was a mistake, and the implementation was left over.